cpp/src/quantiles/groupby.cu

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "groupby.hpp"
+
+#include <copying/scatter.hpp>
+#include <table/legacy/device_table.cuh>
+#include <table/legacy/device_table_row_operators.cuh>
+#include <bitmask/legacy/bit_mask.cuh>
+#include <utilities/column_utils.hpp>
+#include <utilities/cuda_utils.hpp>
+
+#include <cudf/copying.hpp>
+
+#include <thrust/scan.h>
+#include <thrust/binary_search.h>
+#include <thrust/unique.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/discard_iterator.h>
+
+#include <algorithm>
+#include <tuple>
+#include <numeric>
+
+namespace {
+
+template <bool nullable = true>
+struct transform_row_eq_comparator {
+  row_equality_comparator<nullable> cmp;
+  gdf_size_type* sorted_order;
+
+  CUDA_DEVICE_CALLABLE
+  bool operator() (gdf_size_type lhs, gdf_size_type rhs) {
+    return cmp(sorted_order[lhs], sorted_order[rhs]);
+  }
+};
+
+struct permutation_label_setter {
+  gdf_size_type* group_labels_ptr;
+  gdf_size_type* group_ids_ptr;
+
+  CUDA_DEVICE_CALLABLE
+  void operator() (gdf_size_type i) { 
+    group_labels_ptr[group_ids_ptr[i]] = 1;
+  }
+};
+
+} // namespace anonymous
+
+
+namespace cudf {
+
+namespace detail {
+
+void groupby::set_key_sort_order() {
```

Comment:
@aocsa correct me if I am wrong, but this is being called before we do the aggregate, right? If the keys are already sorted, say by prior call to `gdf_order_by`, could we skip this all together?

There are some cases in Spark where data may already be sorted, and so doing this here also looks like this could add extra work.

Reply:
Yes. In fact, this was the original plan, to have an arg that states if the column is already sorted, but left out until we have better argument passing.

Reply:
> @aocsa correct me if I am wrong, but this is being called before we do the aggregate, right? If the keys are already sorted, say by prior call to `gdf_order_by`, could we skip this all together?
> 
> There are some cases in Spark where data may already be sorted, and so doing this here also looks like this could add extra work.

Could use this in Python land as well, we have a lazily computed / memoized property of whether a column is sorted or not that we can use really effectively here.

Reply:
@abellina did you mean `gdf_order_by`? If I understand you correctly, you want to specify whether the key/value columns should be sorted ascending/descending while calculating groups? And separately, you want the ability to consider `NULLS_AS_SMALLEST` while calculating groups? We can do that. Can you provide examples of aggregations where these would give a different result. I'm not doubting, it'd just be easier as I don't have a spark environment.

Reply:
That (ascending, nulls last) **IS** the current default implementation here.

Reply:
Because of @jrhemstad 's review on the group quantiles PR, I am turning all of the members of groupby object into optionals. Also, I think that the whole Options struct should be passed into the groupby object and then let it take care of things.